pkg/jobs/hypershift/hypershift.go

```diff
@@ -0,0 +1,533 @@
+
+		klog.V(2).Infof("Patching %v %v in namespace %v ✓", resourceType.Resource, clusterName, namespace)
+		_, err := dc.Resource(resourceType).Namespace(namespace).Patch(
+			context.TODO(), clusterName, types.JSONPatchType, patchInBytes, v1.PatchOptions{})
+		if err != nil {
+			return err
+		}
+		log.Println("Updated " + resourceType.Resource + " ✓")
+		return nil
+	})
+
+	return err
+}
+
+func validateUpgradeVersion(
+	client clientv1.Client,
+	clusterName string,
+	curator *clustercuratorv1.ClusterCurator,
+	desiredUpdate string) error {
+	managedClusterInfo := managedclusterinfov1beta1.ManagedClusterInfo{}
+	if err := client.Get(context.TODO(), types.NamespacedName{
+		Namespace: clusterName,
+		Name:      clusterName,
+	}, &managedClusterInfo); err != nil {
+		return err
+	}
+
+	if managedClusterInfo.Status.DistributionInfo.OCP.Version == desiredUpdate {
+		return errors.New("Cannot upgrade to the same version")
+	}
+
+	desiredUpdateSplit := strings.Split(desiredUpdate, ".")
+	currentVersionSplit := strings.Split(managedClusterInfo.Status.DistributionInfo.OCP.Version, ".")
+	desiredUpdateMajor, _ := strconv.Atoi(desiredUpdateSplit[0])
+	desiredUpdateMinor, _ := strconv.Atoi(desiredUpdateSplit[1])
+	desiredUpdatePatch, _ := strconv.Atoi(desiredUpdateSplit[2])
+	currentVersionMajor, _ := strconv.Atoi(currentVersionSplit[0])
+	currentVersionMinor, _ := strconv.Atoi(currentVersionSplit[1])
+	currentVersionPatch, _ := strconv.Atoi(currentVersionSplit[2])
+
+	lowerDesiredUpdateVersionErrMsg := "Cannot upgrade to an earlier version"
+
+	if desiredUpdateMajor < currentVersionMajor {
+		return errors.New(lowerDesiredUpdateVersionErrMsg)
+	}
+
+	if desiredUpdateMajor == currentVersionMajor && desiredUpdateMinor < currentVersionMinor {
+		return errors.New(lowerDesiredUpdateVersionErrMsg)
+	}
+
+	if desiredUpdateMajor == currentVersionMajor &&
+		desiredUpdateMinor == currentVersionMinor &&
+		desiredUpdatePatch < currentVersionPatch {
+		return errors.New(lowerDesiredUpdateVersionErrMsg)
+	}
+
+	return nil
+}
+
+func DestroyHostedCluster(dc dynamic.Interface, clusterName string, namespace string) error {
+	klog.V(0).Infof("Deleting Hosted Cluster for %v in namespace %v\n", clusterName, namespace)
+
+	_, err := dc.Resource(utils.HCGVR).Namespace(namespace).Get(context.TODO(), clusterName, v1.GetOptions{})
+
+	if err != nil && k8serrors.IsNotFound(err) {
+		klog.Warning("Could not retreive hosted cluster " + clusterName + " may have already been deleted")
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	err = dc.Resource(utils.HCGVR).Namespace(namespace).Delete(context.TODO(), clusterName, v1.DeleteOptions{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func DetachAndMonitor(dc dynamic.Interface, clusterName string, curator *clustercuratorv1.ClusterCurator) error {
+	var mcGVR = schema.GroupVersionResource{
+		Group:    "cluster.open-cluster-management.io",
+		Version:  "v1",
+		Resource: "managedclusters",
```

Comment:
We do import the `open-cluster-management.io/api` so you can just use the regular client in this function if you want but it's not a big deal. Up to you.